### ISSUE TITLE
Bug: [restrict-template-expressions] incorrect inferred type after ternary operator

### ISSUE THREAD

**Issue by taozhou-glean** [typescript-eslint/typescript-eslint#6941]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=FAFwngDgpgBAgjAvDA3sGMBGB+AXK9DGAbQGsox8BnEAJwEsA7AcwF19GBXAW0ylpgAfGJ0YATKADMmUMQG5CAXwXLgwKAA8IAe1ogYAY22MaMANIUkBDHHwAiAIZ2ANIQBC9g3eCKYDqobGNGqaOnow4NDmlsgAFJFQ2pLRYACUZBRJEZCJyRZgrGpGJvoO+AjIKKoOAHQ4NcQGrDDYMAAGIFCmACQotfWNrIptMPgA5GMKwMWmYlb92A1N00GltMxWc60dXfq9YsOjMBNyQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

(can not repro in the playground as it complains about `c` being not defined, and the issue is gone once c is defined)

### Repro Code

```typescript
test case:

type A = {
  b?: {
    [key: string]: number | undefined;
  };
};
const a: A = {};
const arg = a.b?.[c] ? \`test \${a.b?.[c]}\` : '';
```

expect: no error

actual:
```
● restrict-template-expressions › valid ›
    Message:
      Should have no errors but had 1: [
      {
        ruleId: 'restrict-template-expressions',
        severity: 1,
        message: 'Invalid type "number | undefined" of template literal expression.',
        line: 8,
        column: 38,
        nodeType: 'ChainExpression',
        messageId: 'invalidType',
        endLine: 8,
        endColumn: 46
      }
    ]
```

while
```
type A = {
  b?: {
    [key: string]: number | undefined;
  };
};
const a: A = {};
const d = a.b?.[c]
const arg = d ? \`test \${d}\` : '';
```

can pass without any issue, in both cases, the `c` is not defined, and once `c` is defined, then both tests pass.

```


### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/restrict-template-expressions": ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

listed above

### Actual Result

listed above

### Additional Info

this is the minimum version extracted from real code, I can not repro this on playground as once I define `c`, then all errors are gone, similarly in the test code, once I define `const c = "asd"`, then both test cases passed, without that, only one can pass in `yarn test`

**Comment by taozhou-glean** [comment]
I think the issue is in
```
 const expressionType = util.getConstrainedTypeAtLocation(
            typeChecker,
            service.esTreeNodeToTSNodeMap.get(expression),
          );
```

as if I print the `intrinsicName` type for both tests, i get two different types:

```
// without temp var `d`
    ===== undefined

// with temp var `d`
    ===== number
```

**Comment by Josh-Cena** [comment]
Please see https://github.com/microsoft/TypeScript/issues/10530 and it's 300 duplicates: type narrowing only happens with literal property access, not dynamic access. This is a limitation of TypeScript itself.
